### PR TITLE
Refactor embeddings

### PR DESCRIPTION
**Pull request by aaprasad**
* refactor how we initialize and use embeddings:
  * getting rid of `**kwargs`
  * initialize in params in `__init__`
  * introduce `type` and `mode` params to initialize correct embedding
  * wrap embedding computation function in `forward`
  * adapt embedding init based on refactor, and separate embeddings into pos and temp in transformer class
  * use embedding forward to compute embedding and simplify embedding computation
 
* get rid of redundant args in transformer and get rid of redundant args in GTR class

## Summary by CodeRabbit

- **Refactor**
	- Enhanced flexibility and clarity in model initialization parameters.
	- Improved handling and computation of embeddings, including updates based on metadata.

- **Tests**
	- Introduced and updated test functions for various model configurations and logic.

- **Documentation**
	- Updated configuration files to improve structure and clarity, separating modes and updating keys/values.

### CODE REVIEW

File: biogtr/models/embedding.py [hunk 1]
```diff
@@ -144,13 +199,11 @@ def _learned_pos_embedding(
                 dim=1,
             )
 
-        l, r, lw, rw = self._compute_weights(xywh, self.learn_pos_emb_num)
+        l, r, lw, rw = self._compute_weights(xywh, self.emb_num)
```
[the remark this answers is not in the document]
Comment (aaprasad) [reply biogtr/models/embedding.py]: done in [79810d7](https://github.com/talmolab/biogtr/pull/29/commits/79810d74758c9e21d999cccd0e901192ff6633df)

File: biogtr/models/embedding.py [hunk 2]
```diff
@@ -34,42 +139,15 @@
...
-    def _sine_box_embedding(
-        self,
-        boxes,
-        features: int = 512,
-        temperature: int = 10000,
-        scale: float = None,
-        normalize: bool = False,
-        **kwargs,
-    ) -> torch.Tensor:
+    def _sine_box_embedding(self, boxes: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (aaprasad) [reply biogtr/models/embedding.py:142]: This is not the correct change to make

File: biogtr/models/embedding.py [hunk 3]
```diff
@@ -144,92 +199,67 @@
...
+        left_ind = data.clamp(min=0, max=self.emb_num - 1).long()  # N x 4
+        right_ind = (left_ind + 1).clamp(min=0, max=self.emb_num - 1).long()  # N x 4
 
-        left_weight = data - left_index.float()  # N x 4
+        left_weight = data - left_ind.float()  # N x 4
 
         right_weight = 1.0 - left_weight
 
-        return left_index, right_index, left_weight, right_weight
+        return left_ind, right_ind, left_weight, right_weight
```
[the remark this answers is not in the document]
Comment (aaprasad) [reply biogtr/models/embedding.py:265]: Done in [79810d7](https://github.com/talmolab/biogtr/pull/29/commits/79810d74758c9e21d999cccd0e901192ff6633df)

File: biogtr/models/embedding.py [hunk 4]
```diff
@@ -34,42 +139,15 @@ def _torch_int_div(
...
-        }
-
-        self.features = params["features"]
-        self.temperature = params["temperature"]
-        self.scale = params["scale"]
-        self.normalize = params["normalize"]
-
         if self.scale is not None and self.normalize is False:
             raise ValueError("normalize should be True if scale is passed")
         if self.scale is None:
```
Comment (talmo) [inline biogtr/models/embedding.py:153]: Prefer to set the default value of `self.scale` in a single place (either here or in the constructor)

File: biogtr/models/embedding.py [hunk 5]
```diff
@@ -34,42 +139,15 @@ def _torch_int_div(
...
-        temperature: int = 10000,
-        scale: float = None,
-        normalize: bool = False,
-        **kwargs,
-    ) -> torch.Tensor:
+    def _sine_box_embedding(self, boxes: torch.Tensor) -> torch.Tensor:
         """Compute sine positional embeddings for boxes using given parameters.
 
         Args:
             boxes: the input boxes.
```
Comment (talmo) [inline biogtr/models/embedding.py:146]: Document the expected shapes of `boxes` since there's some shape handling below (e.g., for rank 2 inputs)

File: biogtr/models/embedding.py [hunk 6]
```diff
@@ -81,9 +159,11 @@ def _sine_box_embedding(
         if self.normalize:
             boxes = boxes / (boxes[:, -1:] + 1e-6) * self.scale
 
-        dim_t = torch.arange(self.features, dtype=torch.float32)
+        dim_t = torch.arange(self.features // 4, dtype=torch.float32)
 
-        dim_t = self.temperature ** (2 * self._torch_int_div(dim_t, 2) / self.features)
+        dim_t = self.temperature ** (
+            2 * self._torch_int_div(dim_t, 2) / self.features // 4
```
Comment (talmo) [inline biogtr/models/embedding.py:165]: For clarity on the order of operations:

```suggestion
            (2 * self._torch_int_div(dim_t, 2)) / (self.features // 4)
```

File: biogtr/models/embedding.py [hunk 7]
```diff
@@ -97,41 +177,16 @@ def _sine_box_embedding(
...
-        features: int = 1024,
-        learn_pos_emb_num: int = 16,
-        over_boxes: bool = True,
-        **kwargs,
-    ) -> torch.Tensor:
+    def _learned_pos_embedding(self, boxes: torch.Tensor) -> torch.Tensor:
         """Compute learned positional embeddings for boxes using given parameters.
 
         Args:
             boxes: the input boxes.
```
Comment (talmo) [inline biogtr/models/embedding.py:184]: document shape of `boxes`

File: biogtr/models/embedding.py [hunk 8]
```diff
@@ -144,92 +199,67 @@ def _learned_pos_embedding(
                 dim=1,
             )
 
-        l, r, lw, rw = self._compute_weights(xywh, self.learn_pos_emb_num)
+        left_ind, right_ind, left_weight, right_weight = self._compute_weights(xywh)
 
         f = pos_lookup.weight.shape[1]
```
Comment (talmo) [inline biogtr/models/embedding.py:204]: is this `D * 4`, `d`, `features`, `features // 4` or what?

File: biogtr/models/embedding.py [hunk 9]
```diff
@@ -144,92 +199,67 @@ def _learned_pos_embedding(
...
-        times: torch.Tensor,
-        features: int = 1024,
-        learn_temp_emb_num: int = 16,
-        **kwargs,
-    ) -> torch.Tensor:
+    def _learned_temp_embedding(self, times: torch.Tensor) -> torch.Tensor:
         """Compute learned temporal embeddings for times using given parameters.
 
         Args:
             times: the input times.
```
Comment (talmo) [inline biogtr/models/embedding.py:226]: document shape and contents! integer inds? relative? normalized?

File: biogtr/models/global_tracking_transformer.py [hunk 10]
```diff
@@ -19,15 +19,13 @@ def __init__(
...
-        dim_feedforward: int = 1024,
         dropout: int = 0.1,
         activation: str = "relu",
         return_intermediate_dec: bool = False,
-        feature_dim_attn_head: int = 1024,
         norm: bool = False,
         num_layers_attn_head: int = 2,
         dropout_attn_head: int = 0.1,
-        embedding_meta: dict = None,
+        embedding_meta: dict = {"pos": None, "temp": None},
```
Comment (talmo) [inline biogtr/models/global_tracking_transformer.py:28]: No mutables as defaults! Set it to `None` and initialize it as the dictionary inside the constructor. Same for `encoder_cfg`. If you change those objects after instance creation, they'll change the defaults for all future instances of the class.

File: biogtr/models/global_tracking_transformer.py [hunk 11]
```diff
@@ -42,40 +40,21 @@ def __init__(
...
-                        }
-                    }
-                Note: Embedding features are handled directly in the forward
-                pass for each case. Overriding the features through kwargs will
-                likely throw errors due to incorrect tensor shapes.
+                that no positional embeddings should be used. To use the positional embeddings
+                pass in a dictionary containing a "pos" and "temp" key with subdictionaries for correct parameters ie:
+                {"pos": {'mode': 'learned', 'emb_num': 16, 'over_boxes: 'True'},
+                "temp": {'mode': 'learned', 'emb_num': 16}
+                }. See `embedding.py` for more information on correct arguments for initialization.
```
Comment (talmo) [inline biogtr/models/global_tracking_transformer.py:57]: Refer to a valid symbol (module/class/function name), not files.

File: biogtr/models/embedding.py [hunk 12]
```diff
@@ -13,12 +13,117 @@ class Embedding(torch.nn.Module):
     Used for both learned and fixed embeddings.
     """
 
-    def __init__(self):
-        """Initialize embeddings."""
+    EMB_TYPES = {
+        "temp": {},
+        "pos": {"over_boxes"},
+        "": {},
```
Comment (talmo) [inline biogtr/models/embedding.py:19]: call this `"null"` or `"zeros"` or something that's more descriptive than an empty string so the behavior is a bit clearer from the arguments.

**Review by talmo — CHANGES_REQUESTED** [review]

**Review by talmo — APPROVED** [review]
